hbase-client/src/main/java/org/apache/hadoop/hbase/client/AdminOverAsyncAdmin.java

```diff
@@ -1057,4 +1057,9 @@ public void updateRSGroupConfig(String groupName, Map<String, String> configurat
       throws IOException {
     get(admin.updateRSGroupConfig(groupName, configuration));
   }
+
+  @Override
+  public List<BalancerDecisionRecords> getBalancerDecisions() throws IOException {
```

Comment:
Something to consider: Rather than adding new API for every ringbuffer backed type, since the ringbuffers are named, can we just have one API that retrieves records from a buffer specified by name? 

E.g.

public List<LogEntry> getLogEntries(String name)

Then, LogEntry is a generic type capable of accepting any protobuf encoding. Then, we derive new types from LogEntry such as BalancerDecision. Have a static method in RingEntry for instantiating the subclasses based on what type is communicated by the protobuf. 

If _LogEntry_ is too generic a name, consider _RingEntry_. (I'm not the best at naming, maybe someone else has a better idea...)

It is a lot easier to add or remove specialized classes as these things evolve than add or remove methods from public/stable admin APIs.

Reply:
This would mean you'd have to go back and modify any admin API related to the slow log, which is fine, and desirable (if you accept the premise of this feedback)

Reply:
If we have this type of API, it should be possible to provide a limit, e.g.

public List getLogEntries(String name, int limit);

so that a client that is memory constrained (or wants to be frugal) doesn't have to worry about invoking this and maybe getting back a list of 5000 entries or whatever, by passing in a limit of 100, or 10, or ...

Reply:
Definitely, adding new APIs not desirable, however the reason why I had to choose this way: 

1. slowLog released in 2.3 already.
2. slowLog is in each RS and the API has flexibility to choose serverName to get records from (filter by serverName) whereas balancer is HMaster queue meaning activeMasterStub is to be identified by client, hence no need of serverName filter.
3. client has to identify whom to make a call - HMaster or RegionServer (or limited no of RS) based on use-case, not a big deal but maybe too much logic for client interface?
4. genericAPI might need to include abstract request and response classes, and respective use-cases to ensure if filter is required at server side (not a big deal)

For 1), I believe trunk and 2.4.0 both can have generic API end-point.
2) and 3) seem bit of headache for single API.

Reply:
I agree, that's the trade off here. Some pain now for flexibility later, or not much pain and no flexibility now or later. New APIs for each new type. 

Either decision is not wrong or right. 

A benefit to API per ring type is pressure from API compat concerns and method proliferation will put a lot of pressure on new potential use cases for this mechanism, will set a really high bar. Maybe too high, hence my concerns.

Reply:
So just to be clear, please take some time to think about this, but it is not a request for change per se.

Reply:
Ok, but if you dont' accept the majority of this feedback, which would be fine, we at least need the limit parameter. And also add one for the slow log if there isn't one there.

Reply:
I agree method proliferation is a concern, something to consider and once we come to a decision, there is no way back it seems.
For limit param, we have limit in slowLog use-case because client is calling every single RS. Here, since it's just HMaster, the real limit should be kept in HMaster's ring buffer size. However, memory constraint makes limiting useful even for single server.